tests/kafkatest/tests/streams/streams_relational_smoke_test.py

```diff
@@ -0,0 +1,123 @@
+                   "log4j": self.LOG4J_CONFIG_FILE,
+                   "kafka_run_class": self.path.script("kafka-run-class.sh", node),
+                   "mode": self.mode,
+                   "nodeId": self.nodeId,
+                   "kafka": self.kafka.bootstrap_servers(),
+                   "state_dir": self.PERSISTENT_ROOT,
+                   "stdout": self.STDOUT_FILE,
+                   "stderr": self.STDERR_FILE,
+                   "pidfile": self.PID_FILE
+               }
+
+    def start_node(self, node):
+        node.account.mkdirs(self.PERSISTENT_ROOT)
+        node.account.create_file(self.LOG4J_CONFIG_FILE,
+                                 self.render("log4j_template.properties", log_file=self.LOG_FILE))
+
+        self.logger.info("Starting process on " + str(node.account))
+        node.account.ssh(self.start_cmd(node))
+
+        if len(self.pids(node)) == 0:
+            raise RuntimeError("No process ids recorded")
+
+    def await_command(self, command):
+        wait_until(lambda: self.node.account.ssh(command, allow_fail=True),
+                   timeout_sec=60,
+                   err_msg="Command [%s] never passed in the timeout"
+                   )
+
+
+class StreamsRelationalSmokeTest(KafkaTest):
+    """
+    Simple test of Kafka Streams.
+    """
+
+    def __init__(self, test_context):
+        super(StreamsRelationalSmokeTest, self).__init__(test_context, num_zk=1, num_brokers=3, topics={
+            'in-article': {'partitions': 3, 'replication-factor': 1},
+            'in-comment': {'partitions': 5, 'replication-factor': 1},
+            'out-augmented-article': {'partitions': 3, 'replication-factor': 1},
+            'out-augmented-comment': {'partitions': 5, 'replication-factor': 1}
+        })
+        self.test_context = test_context
+
+    @cluster(num_nodes=8)
+    @matrix(crash=[False, True])
+    def test_streams(self, crash):
+        driver = StreamsRelationalSmokeTestService(self.test_context, self.kafka, "driver", "ignored")
+
+        LOG_FILE = driver.LOG_FILE  # this is the same for all instaces of the service, so we can just declare a "constant"
+
+        processor1 = StreamsRelationalSmokeTestService(self.test_context, self.kafka, "application", "processor1")
+        processor2 = StreamsRelationalSmokeTestService(self.test_context, self.kafka, "application", "processor2")
+
+        processor1.start()
+        processor2.start()
+
+        processor1.await_command("grep -q 'Streams has started' %s" % LOG_FILE)
+        processor2.await_command("grep -q 'Streams has started' %s" % LOG_FILE)
+
+        driver.start()
+
+        # wait for at least one output of stream processing before injecting failures
+        driver.await_command("grep -q 'Consumed first Augmented' %s" % LOG_FILE)
+
+        processor1.stop_nodes(not crash)
+
+        processor3 = StreamsRelationalSmokeTestService(self.test_context, self.kafka, "application", "processor3")
```

Comment:
In other tests we start all three processors, and the rolling bounce them; here we start 2, shutdown 1, and start a new one, and shutdown another. Is it intentional and what's the motivation?

Reply:
I wanted to make sure that the computation survives state shuffling as a result of processor 3 having to recover all of processor 1's state, and then subsequently all of processor 2's state.

Conversely, if we just do a rolling bounce, then instances would still be able to make use of the state from before.

Do you think this makes sense, or is it just fuzzy thinking on my part?

Reply:
Okay I think this makes sense, let's just follow this pattern then.